### PR TITLE
Allow explicitly enabling or disabling tablets when creating a new keyspace

### PR DESCRIPTION
**Pull request by bhalevy**
Separate the configuration for enabling the tablets feature from the enablement of tablets when creating new keyspaces.

This change always enables the TABLETS cluster feature and the tablets logic respectively.

The `enable_tablets` config option just controls whether tablets are enabled or disabled by default for new keyspaces.

If `enable_tablets` is set to `true`, tablets can be disabled using `CREATE KEYSPACE WITH tablets = { 'enabled': false }` as it is today.

If `enable_tablets` is set to `false`, tablets can be enabled using `CREATE KEYSPACE WITH tablets = { 'enabled': true }`.

The motivation for this change is to simplify the user experience of using tablets by setting the default for new keyspaces to false amd allowing the user to simply opt-in by using tablets = {enabled: true }.
This is not pissible today.
The user has to enable tablets by default for all new keyspaces (that use the NetworkTopologyStrategy) and then actively opt-out to use vnodes.

* Not required to be backported to OSS versions.  May be backported to specific enterprise versions

### CODE REVIEW

**Comment by mykaul** [comment]
We probably should have also 'always' and 'never'. 'Always' is important as we may want tools to be able to assume this is a tablets-only setup, for example.

**Comment by bhalevy** [comment]
> We probably should have also 'always' and 'never'. 'Always' is important as we may want tools to be able to assume this is a tablets-only setup, for example.

@mykaul can you be more specific please?
How do you see `always` and `never` used?

**Comment by mykaul** [comment]
> > We probably should have also 'always' and 'never'. 'Always' is important as we may want tools to be able to assume this is a tablets-only setup, for example.
> 
> @mykaul can you be more specific please? How do you see `always` and `never` used?

Always: default in some new deployments - where we only support Tablets
Never: default when upgrading an existing deployment that does not have Tablets yet
(could be different values set in enterprise vs. open source, cloud vs. enterprise, etc.)

**Comment by avikivity** [comment]
No one will understand a configuration file that has enable_tablets and enable_tablet_feature. Why is this so complicated? flags for enabling other flags?

**Comment by bhalevy** [comment]
> No one will understand a configuration file that has enable_tablets and enable_tablet_feature. Why is this so complicated? flags for enabling other flags?

`enable_tablets_feature` can be hidden from the user or even dropped so that the feature would always be enabled.

**Comment by bhalevy** [comment]
Summary of meeting we held today with @tzach, @avikivity, and @dani-tweig:
- Keep only one option: `enable_tablets`
- Rename the values to `opt_in` / `opt_out`
  - rather than `false` / `true`, respectively
  - old values will be supported for backward compatibility
  - but documentation will be modified to reflect the change
- Tablets logic would be enabled by default with no option to disable it

**Comment by dani-tweig** [comment]
> Rename the values to `opt_in` / `opt_out`
> * rather than `false` / `true`, respectively
> * old values will be supported for backward compatibility
> * but documentation will be modified to reflect the change

These settings refer to creating a new keyspace.
`opt-in` / `false` - means that new keyspaces will be created with vnodes, unless you specify otherwise in CQL.
`opt-out` / `true` - means that new keyspaces will be created with tablets, unless you specify otherwise in CQL.

File: docs/architecture/tablets.rst
Comment (bhalevy) [reply docs/architecture/tablets.rst]: docs revised in next version

File: docs/architecture/tablets.rst
Comment (bhalevy) [reply docs/architecture/tablets.rst]: docs revised in next version

**Comment by bhalevy** [comment]
Required dtest changes: https://github.com/scylladb/scylla-dtest/pull/4998

**Comment by bhalevy** [comment]
> #### 🔴 CI State: FAILURE
> ❌ - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/11836/)
> 
> #### Build Details:
> * Duration: 11 min
> * Builder: spider2.cloudius-systems.com

looks environmental;
```
[2024-09-23T18:09:02.260Z] pip._vendor.urllib3.exceptions.ReadTimeoutError: 
```

respinning

**Comment by bhalevy** [comment]
> @bhalevy @dani-tweig @mykaul @tzach I love this new implementationa and I'll re-review the docs, but before that, I'd like to add my feedback on configuring the option.
> 
> As a user, I opened the _scylla.yaml_ file before I resorted to the docs. What I saw in _scylla.yaml_ was: `enable_tablets: opt-out` This suggested to me that I my cluster is configured to opt-out from enabling the tablets by default. The descirpiton is correct (_When set to "opt-out", newly created keyspaces will have tablets enabled by default._), but it doesn't help, as "opt-out" and "enabled by default" are seemingly opposite, so the entire configuration is counterintuitive and confusing.
> 
> The configuration you proposed makes sense in the light of the discussion above. However, the user is not aware of that dicussion. To the user `enable_tablets: opt-out` will mean "tablets are disabled by default, because you opted out", not "tablets are enabled, but you can opt out".
> 
> The values are related to what the user can do **next**, not what the config option acutally does, that's the core of the confusion.
> 
> IMHO, the configuration should stay as they were, because they were clear: `enable_tablets: true` - new keyspaces will be created with tablets, unless you specify otherwise in CQL. `enable_tablets: false` - new keyspaces will be created with vnodes, unless you specify otherwise in CQL.
> 
> And in the documentation (and the description in scylla.yaml) we can add the information that "tablets are enabled/disabled by default for your keyspace, but you can opt out/in by explicitly configuring the `tablets = {'enabled': false}`/`tablets = {'enabled': true}` option when creating a keyspace.

@avikivity are you ok with that?
I'm inclined to agree with @annastuchlik that `tablets_enabled: false/true` is clearer and simpler to implement than opt-in/opt-oit values (see last patch in this series)

**Comment by bhalevy** [comment]
@avikivity ping

**Comment by mykaul** [comment]
@bhalevy - please resolve conflicts and let's try to get it in soonish.

**Comment by bhalevy** [comment]
In v3 ([b2de063](https://github.com/scylladb/scylladb/commit/b2de063f416b0c998cd6eae0f0b0fd9a2348503e)):
- rebased
- dropped "db: config: enable_tablets: accept opt_in_out strings"

**Comment by bhalevy** [comment]
In v4 ([a285d49](https://github.com/scylladb/scylladb/commit/a285d496438d46f79de1508b33893b60b19cb0e6)):
- treewide: always enable tablets logic
  - adjusted test_tablets
- data_dictionary: keyspace_metadata::describe: print tablets enabled also when defaulted

Still need to adjust the dtests that this change breaks.

**Comment by avikivity** [comment]
> Should be backported to all versions supporting tablets, for consistency.

You're making 6.n be consistent with 6.(n-1), at the cost of making 6.(n-1).m inconsistent with 6.(n-1).(m+1). I don't think it's a good idea to make such changes in patch releases.

File: cql3/statements/create_keyspace_statement.cc
Comment (avikivity) [inline cql3/statements/create_keyspace_statement.cc:96]: > so that even when set to `false`, tablets would
> still be enabled as a feature, so that a following

This certainly shouldn't be backported, it's a large change.

**Comment by avikivity** [comment]
> If enable_tablets is set to opt-out (or true), tablets can be disabled using CREATE KEYSPACE WITH tablets = { 'enabled': false } as it is today.


Didn't see an "opt-in" string.

**Comment by bhalevy** [comment]
> > If enable_tablets is set to opt-out (or true), tablets can be disabled using CREATE KEYSPACE WITH tablets = { 'enabled': false } as it is today.
> 
> Didn't see an "opt-in" string.

I dropped this patch from the recent version based on @annastuchlik's feedback in https://github.com/scylladb/scylladb/pull/20729#issuecomment-2374037239

**Comment by avikivity** [comment]
> > > If enable_tablets is set to opt-out (or true), tablets can be disabled using CREATE KEYSPACE WITH tablets = { 'enabled': false } as it is today.
> > 
> > 
> > Didn't see an "opt-in" string.
> 
> I dropped this patch from the recent version based on @annastuchlik's feedback in [#20729 (comment)](https://github.com/scylladb/scylladb/pull/20729#issuecomment-2374037239)

The changelog still refer to those strings.

**Comment by bhalevy** [comment]
> > > > If enable_tablets is set to opt-out (or true), tablets can be disabled using CREATE KEYSPACE WITH tablets = { 'enabled': false } as it is today.
> > > 
> > > 
> > > Didn't see an "opt-in" string.
> > 
> > 
> > I dropped this patch from the recent version based on @annastuchlik's feedback in [#20729 (comment)](https://github.com/scylladb/scylladb/pull/20729#issuecomment-2374037239)
> 
> The changelog still refer to those strings.

sorry, I"ll sanitize it

**Comment by bhalevy** [comment]
In v5 ([52dcc8b](https://github.com/scylladb/scylladb/commit/52dcc8b32526256a69f5097b3eb9780835f94f57)):
- Edited subject and description of this PR, and commit messages, not mentioning `opt-in` or `opt-out` configuration.
- cleaned up and renamed added unit tests
- removed backport labels

**Comment by kostja** [comment]
@bhalevy sorry since I am late in this discussion, we had a similar discussion, but it impacted the keyspaces work by @ptrsmrn so we had a chat about it with him and @xemul as well.

I think enable_tablets: is a good name as it is and is should stay as it is, enable/disable the tablets feature.

Then we could have a separate setting, something like partitioning_strategies: which can accept a list of allowed partitioning strategies. The first item in the list woudl be the default.

Another matter is quietly downgrading the partitioning strategy to vnode for Simple Replication Strategy. This seems to be a pure evil and should be disabled altogether.

**Comment by bhalevy** [comment]
> @bhalevy sorry since I am late in this discussion, we had a similar discussion, but it impacted the keyspaces work by @ptrsmrn so we had a chat about it with him and @xemul as well.
> 
> I think enable_tablets: is a good name as it is and is should stay as it is, enable/disable the tablets feature.

The name is kept in this PR.  But the semantics were extended.  In this branch `enable_tablets` controls the default for newly created keyspaces.  While, unlike today, the default can be overridden by the user using the `tablets = {`enabled`: true/false}` replication option.

> Then we could have a separate setting, something like partitioning_strategies: which can accept a list of allowed partitioning strategies. The first item in the list woudl be the default.

https://github.com/scylladb/scylladb/issues/14208 / https://github.com/scylladb/scylladb/issues/16028
Cc @nyh

> Another matter is quietly downgrading the partitioning strategy to vnode for Simple Replication Strategy. This seems to be a pure evil and should be disabled altogether.

I agree.  With this PR it makes even more sense to refuse using SimpleStrategy when tablets are enabled, either by default, or using the tablets option to CREATE KEYSPACE.
So to use SimpleStrategy when tablets are enabled by default one will have to disable tablets explicitly.
This can be done orthogonally.
Can you please open an issue about that?

**Comment by bhalevy** [comment]
Unrelated CI failure: https://github.com/scylladb/scylladb/issues/21034
Respinned

**Comment by bhalevy** [comment]
> > No one will understand a configuration file that has enable_tablets and enable_tablet_feature. Why is this so complicated? flags for enabling other flags?
> 
> I agree. Two more thoughts:
> 
> 1. Why do we even need the "--enable_tablet_feature" option? If a user does not use tablets on any keyspace, the tablets shouldn't change anything. If we're worried that maybe the tablet code can cause bugs even in workloads that never use it, maybe we should call this option "--disable-tablets" - and it will only be used in emergencies in the field, not by ordinary users.

@nyh, your comment is based on outdated information.  The recent versions of the change dropped the proposed `enable_tablet_feature` option and just always enable this feature by default (and use it based on the cluster feature bit).

> 2. I think the new "--enable-tablets" option controlling the default use of tablets in CQL keyspaces (sadly not Alternator with LWT...) is misnamed, because as you noted, even if enable-tablets=false you can still use tablets just fine - you just need to ask for it per keyspace. A name like --default-tablets or even (yuck) --cql-default-tablets would be more accurate.

**Comment by kostja** [comment]
> > Then we could have a separate setting, something like partitioning_strategies: which can accept a list of allowed partitioning strategies. The first item in the list woudl be the default.
> 
> #14208 / #16028 Cc @nyh
> 

What I am advocating is that Tablets vs VNodes is not a question of Replication Strategy, it's Partitioning Strategy. 

So the above tickets which guardrail harmful replication strategies, and we discuss here the partitioning strategy. Not every replication strategy works with every partitioning strategy - this is OK.

**Comment by bhalevy** [comment]
@kostja IIUC, that's not how the partitioning concept was defined by Cassandra.
See https://docs.datastax.com/en/cassandra-oss/3.0/cassandra/architecture/archPartitionerAbout.html
In this model partitioning is about mapping the partition keys to tokens.
And we're talking about how mapping the token ownership, i.e. which are the token replicas.
It looks like in their model, this is part of data distribution (https://docs.datastax.com/en/cassandra-oss/3.0/cassandra/architecture/archDataDistributeDistribute.html?hl=vnodes)
so maybe a better name would be data_distribution_strategy.

That said, do we have plans for other data distribution strategies in the future to require a (extensible) list?
We want to deprecate vnodes as soon as we can and never look back since maintaining both is very costly (especially once you compound that with every other feature like topology changes, LWT, CDC, counters, MV, plus repair, truncate, backup and restore, and so on) and tablets should be competitive with vnodes so I wouldn't want to encourage supporting vnodes as a formal fallback option for too long.

**Comment by kostja** [comment]
It looks like in their model, this is part of data distribution (https://docs.datastax.com/en/cassandra-oss/3.0/cassandra/architecture/archDataDistributeDistribute.html?hl=vnodes) so maybe a better name would be data_distribution_strategy.

```data_distribution_strategy``` is an excellent name as far as I can tell
> 
> That said, do we have plans for other data distribution strategies in the future to require a (extensible) list? We want to deprecate vnodes as soon as we can and never look back since maintaining both is very costly (especially once you compound that with every other feature like topology changes, LWT, CDC, counters, MV, plus repair, truncate, backup and restore, and so on) and tablets should be competitive with vnodes so I wouldn't want to encourage supporting vnodes as a formal fallback option for too long.

When we deprecate vnodes we can deprecate the variable as well, it's a minor nuisance in such a huge undertaking which will span years anyway.

**Comment by bhalevy** [comment]
> @bhalevy I understand the change, but what is the motivation? Any special use case? I mean - it's working right now, why change it. That being said, `enable_tablets` caused me a headache some time ago.

The motivation is https://github.com/scylladb/scylla-enterprise/issues/4659
We want to simplify the user experience of using tablets by setting the default for new keyspaces to false amd allowing the user to simply opt-in by using `tablets = {`enabled`: true }`.
This is not pissible today.
The user has to enable tablets by default for all new keyspaces (that use the NetworkTopologyStrategy) and then actively opt-out to use vnodes.

**Comment by kbr-scylla** [comment]
> @kostja IIUC, that's not how the partitioning concept was defined by Cassandra.
See https://docs.datastax.com/en/cassandra-oss/3.0/cassandra/architecture/archPartitionerAbout.html
In this model partitioning is about mapping the partition keys to tokens.
And we're talking about the mapping of token ownership, i.e. which nodes own each token's replicas.
It looks like in their model, this is part of data distribution (https://docs.datastax.com/en/cassandra-oss/3.0/cassandra/architecture/archDataDistributeDistribute.html?hl=vnodes)
so maybe a better name would be data_distribution_strategy.

The standard name for mapping keys to different shards of a database in distributed algorithms / distributed database literature is "partitioning". I wouldn't introduce another name just because Cassandra used the word "partitioner" to name a part of the full partitioning strategy.

**Comment by bhalevy** [comment]
Note https://github.com/scylladb/scylladb/issues/21075
tablets require raft topology changes, so they cannot be used with --force-gossip-topology-changes

**Comment by bhalevy** [comment]
> #### 🔴 CI State: FAILURE
> ✅ - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/12355/) ❌ - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/12355/) **The following new/updated tests ran 100 times for each mode:** 🔹 boost/tablets_test 🔹 topology_custom/test_tablets
> 
> #### Failed Tests (1/2444):
> * [test_multidc_alter_tablets_rf.dev.3](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/12355/testReport/junit/%28root%29/test_tablets/test_multidc_alter_tablets_rf_dev_3) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_multidc_alter_tablets_rf.dev.3)

https://github.com/scylladb/scylladb/issues/21034

**Comment by bhalevy** [comment]
In v6 ([97b43d1](https://github.com/scylladb/scylladb/commit/97b43d1d282730cdd3b9d704cdf91b504e9b1649)):
- feature_service: prevent enabling both tablets and gossip topology changes
  - Fixes #21075
- treewide: always allow tablets keyspaces
  - renamed patch and based on "feature_service: prevent enabling both tablets and gossip topology changes"
  - updates the additional test_tablets_and_gossip_topology_changes_are_incompatible unit test

**Comment by bhalevy** [comment]
In v8 ([01e9e12](https://github.com/scylladb/scylladb/commit/01e9e12ec599fb79a841832aeca3a23bdb778718)):
- test_topology_upgrade* unit tests: disable tablets when forcing gossip topology changes

**Comment by bhalevy** [comment]
> #### 🔴 CI State: FAILURE
> ✅ - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/12379/) ❌ - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/12379/) **The following new/updated tests ran 100 times for each mode:** 🔹 boost/tablets_test 🔹 topology_custom/test_tablets 🔹 topology_experimental_raft/test_topology_upgrade 🔹 topology_experimental_raft/test_topology_upgrade_not_stuck_after_recent_removal
> 
> #### Failed Tests (1/7310):
> * [test_tablet_rf_change[down].debug.4](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/12379/testReport/junit/%28root%29/test_tablets/test_tablet_rf_change_down__debug_4) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_tablet_rf_change%5Bdown%5D.debug.4)

Unrelated: https://github.com/scylladb/scylladb/issues/21102
Respinning...

**Comment by bhalevy** [comment]
@avikivity, @kostja, @nyh: any more comments?
We'd like to get this change in ASAP.

File: alternator/executor.cc
Comment (avikivity) [inline alternator/executor.cc:4796]: Does this mean that alternator keyspaces will be using tablets unconditionally from now on?

File: gms/feature_service.cc
Comment (avikivity) [inline gms/feature_service.cc]: Isn't this part of raft recovery? Does this mean we cannot recover raft group 0 if tablets are enabled?

@kbr-scylla

Comment (kbr-scylla) [reply gms/feature_service.cc]: yes

Comment (kbr-scylla) [reply gms/feature_service.cc]: actually:
- force_gossip_topology_changes is not part of raft recovery
- we cannot recover raft group 0 if you have any tablet enabled keyspace

Comment (avikivity) [reply gms/feature_service.cc]: Ok, so this patch isn't making it worse.

But we need to fix recovery. Maybe recovery mode should run without any user keyspaces enabled (and with the front-end ports closed), so there would be no tablets keyspaces.

Comment (kbr-scylla) [reply gms/feature_service.cc]: > Maybe recovery mode should run without any user keyspaces enabled (and with the front-end ports closed), so there would be no tablets keyspaces.

This has nothing to do with the reason why recovery mode is incompatible with tablets.

Please consider the procedure as it is described in our documentation: https://opensource.docs.scylladb.com/stable/troubleshooting/handling-node-failures.html
Screenshot from 2024-10-16 17-55-07

And it should become clear why it's incompatible with tablets:

It's incompatible with tablets because you're performing `nodetool removenode` using gossip-based topology as part of the manual recovery procedure (step 4). This obviously does not perform the necessary tablet migrations/rebuilding since the tablets control plane was only implemented in raft-based topology. So once you removed all the dead nodes in gossip mode, you're left with a tablet mapping that doesn't make any sense (it refers to replicas that no longer exist in the cluster etc.)

Furthermore at the end of recovery procedure you perform upgrade-to-raft-topology again (steps 9, 10), which builds raft-topology state (system.topology table and so on) from scratch based on gossiper state (system.peers table and so on) , and this logic also does not take any leftover previous tablet map into account (e.g. if system.topology referred to some tablet mappings before you started recovery, the newly built system.topology will no longer have any of this information, as there is no code implemented to build it).

Comment (avikivity) [reply gms/feature_service.cc]: I see. Well, we need to fix it.

File: data_dictionary/data_dictionary.cc
Comment (tchaikov) [inline data_dictionary/data_dictionary.cc:419]: printing out if "tablets" is "enabled" if it is enabled breaks the existing tests. see https://jenkins.scylladb.com/job/scylla-master/job/next/8340/testReport/junit/cqlsh_tests.cqlsh_tests/TestCqlsh/Build___x86___dtest_with_topology_changes___test_describe/ . 

because https://github.com/scylladb/scylla-dtest/blob/bbd917ceebf01ad7a0185b5ce9442c0ed2e50a57/cqlsh_tests/cqlsh_tests.py#L754 does not include this option yet in its `CREATE KEYSPACE` statement.

if we want to make this explicit, probably we can update `TestCqlsh.execute()` so that it accepts multiple `expected_output`, if any of them matches, we can consider the test pass? and after this change in scylla lands, we can change `TestCqlsh.execute()`, so that it only accepts the `expected_output`  containing `AND tablets ...`.

**Review by tchaikov — CHANGES_REQUESTED** [review]
this breaks the next promotion test, see https://jenkins.scylladb.com/job/scylla-master/job/next/8340/

**Comment by denesb** [comment]
Dequeued, breaks promotion.

File: data_dictionary/data_dictionary.cc
Comment (bhalevy) [reply data_dictionary/data_dictionary.cc:419]: https://github.com/scylladb/scylla-dtest/pull/5128

**Comment by bhalevy** [comment]
> Dequeued, breaks promotion.

@denesb should be fine now with https://github.com/scylladb/scylla-dtest/pull/5128 merged

**Comment by denesb** [comment]
Conflicts, please rebase.

**Comment by bhalevy** [comment]
[71c1db4](https://github.com/scylladb/scylladb/commit/71c1db47978780371c8f43be39aa896f25d47da2): rebased, conflict resolved.

**Comment by bhalevy** [comment]
> #### 🔴 CI State: FAILURE
> ✅ - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/12683/) ❌ - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/12683/) **The following new/updated tests ran 100 times for each mode:** 🔹 boost/tablets_test 🔹 topology_custom/test_tablets 🔹 topology_experimental_raft/test_topology_upgrade 🔹 topology_experimental_raft/test_topology_upgrade_not_stuck_after_recent_removal
> 
> #### Failed Tests (1/2924):
> * [test_tablet_rf_change[down].debug.3](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/12683/testReport/junit/%28root%29/test_tablets/test_tablet_rf_change_down__debug_3) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_tablet_rf_change%5Bdown%5D.debug.3)
> 
> #### Build Details:
> * Duration: 3 hr 42 min
> * Builder: i-0501771130ecd5013 (m5ad.8xlarge)

https://github.com/scylladb/scylladb/issues/21102
respinning...

**Comment by bhalevy** [comment]
[5498018](https://github.com/scylladb/scylladb/commit/5498018cbeef82ce26e86546e8637c2f0dfe8635): rebased to include fix for https://github.com/scylladb/scylladb/issues/21102 (and rerun CI)

**Comment by bhalevy** [comment]
> #### 🔴 CI State: FAILURE
> ✅ - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/12736/) ✅ - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/12736/) **The following new/updated tests ran 100 times for each mode:** 🔹 boost/tablets_test 🔹 topology_custom/test_tablets 🔹 topology_experimental_raft/test_topology_upgrade 🔹 topology_experimental_raft/test_topology_upgrade_not_stuck_after_recent_removal ✅ - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/12736/) ✅ - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/1350/) ✅ - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/12440/) ✅ - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/4418/) ❌ - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/12736/)
> 
> #### Failed Tests (1/43676):
> * [[empty]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/12736/testReport/junit/boost.memtable_test.test_exceptions_in_flush_on_sstable_open.1.xunit/xml/Tests___Unit_Tests____empty_) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+%5Bempty%5D)
> 
> #### Build Details:
> * Duration: 12 hr
> * Builder: i-054adb36739ca2041 (m5ad.8xlarge)

Weird. This looks like https://github.com/scylladb/scylladb/issues/19753
but it's supposed to be fixed already.
Cc @elcallio

**Comment by bhalevy** [comment]
> > #### 🔴 CI State: FAILURE
> > ✅ - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/12736/) ✅ - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/12736/) **The following new/updated tests ran 100 times for each mode:** 🔹 boost/tablets_test 🔹 topology_custom/test_tablets 🔹 topology_experimental_raft/test_topology_upgrade 🔹 topology_experimental_raft/test_topology_upgrade_not_stuck_after_recent_removal ✅ - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/12736/) ✅ - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/1350/) ✅ - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/12440/) ✅ - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/4418/) ❌ - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/12736/)
> > #### Failed Tests (1/43676):
> > 
> > * [[empty]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/12736/testReport/junit/boost.memtable_test.test_exceptions_in_flush_on_sstable_open.1.xunit/xml/Tests___Unit_Tests____empty_) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+%5Bempty%5D)
> > 
> > #### Build Details:
> > 
> > * Duration: 12 hr
> > * Builder: i-054adb36739ca2041 (m5ad.8xlarge)
> 
> Weird. This looks like #19753 but it's supposed to be fixed already. Cc @elcallio

Reproduced locally.  Decoded stacktrace:
```
[Backtrace #0]
__interceptor_backtrace at /mnt/clang_build/llvm-project-x86_64/compiler-rt/lib/asan/../sanitizer_common/sanitizer_common_interceptors.inc:4358
void seastar::backtrace<seastar::backtrace_buffer::append_backtrace()::{lambda(seastar::frame)#1}>(seastar::backtrace_buffer::append_backtrace()::{lambda(seastar::frame)#1}&&) at ./build/debug/seastar/./seastar/include/seastar/util/backtrace.hh:68
seastar::backtrace_buffer::append_backtrace() at ./build/debug/seastar/./build/debug/seastar/./seastar/src/core/reactor.cc:825
seastar::print_with_backtrace(seastar::backtrace_buffer&, bool) at ./build/debug/seastar/./build/debug/seastar/./seastar/src/core/reactor.cc:858
seastar::print_with_backtrace(char const*, bool) at ./build/debug/seastar/./build/debug/seastar/./seastar/src/core/reactor.cc:870
seastar::sigabrt_action() at ./build/debug/seastar/./build/debug/seastar/./seastar/src/core/reactor.cc:4003
seastar::install_oneshot_signal_handler<6, (void (*)())(&seastar::sigabrt_action)>()::{lambda(int, siginfo_t*, void*)#1}::operator()(int, siginfo_t*, void*) const at ./build/debug/seastar/./build/debug/seastar/./seastar/src/core/reactor.cc:3980
seastar::install_oneshot_signal_handler<6, (void (*)())(&seastar::sigabrt_action)>()::{lambda(int, siginfo_t*, void*)#1}::__invoke(int, siginfo_t*, void*) at ./build/debug/seastar/./build/debug/seastar/./seastar/src/core/reactor.cc:3975
/lib64/libc.so.6: ELF 64-bit LSB shared object, x86-64, version 1 (GNU/Linux), dynamically linked, interpreter /lib64/ld-linux-x86-64.so.2, BuildID[sha1]=77c77fee058b19c6f001cf2cb0371ce3b8341211, for GNU/Linux 3.2.0, not stripped

__GI___sigaction at :?
__pthread_kill_implementation at ??:?
__GI_raise at :?
__GI_abort at :?
std::unexpected() at ??:?
std::rethrow_exception(std::__exception_ptr::exception_ptr) at ??:?
std::terminate() at ??:?
__clang_call_terminate at memtable_test.cc:?
seastar::deferred_stop<seastar::sharded<compaction_manager> >::do_stop() at ././seastar/include/seastar/util/closeable.hh:125
seastar::deferred_stop<seastar::sharded<compaction_manager> >::~deferred_stop() at ././seastar/include/seastar/util/closeable.hh:147
 (inlined by) single_node_cql_env::run_in_thread(std::function<seastar::future<void> (cql_test_env&)>, cql_test_config, std::optional<cql_test_init_configurables>) at ././test/lib/cql_test_env.cc:1063
single_node_cql_env::do_with(std::function<seastar::future<void> (cql_test_env&)>, cql_test_config, std::optional<cql_test_init_configurables>)::{lambda()#1}::operator()() const at ././test/lib/cql_test_env.cc:442
```

**Comment by bhalevy** [comment]
@denesb / @scylladb/dtest-maintainers please consider merging

**Comment by dani-tweig** [comment]
@bhalevy , are you going to backport it to 2024.2.0 directly from master? or first through 6.0?

**Comment by bhalevy** [comment]
> @bhalevy , are you going to backport it to 2024.2.0 directly from master? or first through 6.0?

6.0 is closed for business, no?

Also, we decided to not backport this to oss releases, only to enterprise, so I'll submit the backport directly from master to 2024.2

**Comment by avikivity** [comment]
Reverted in 73b1f66b707682c95c6a9427f00a94c5080c2f91

**Comment by bhalevy** [comment]
> Reverted in [73b1f66](https://github.com/scylladb/scylladb/commit/73b1f66b707682c95c6a9427f00a94c5080c2f91)

@avikivity I'm a bit confused. How does this change affect memtable_test?

**Comment by bhalevy** [comment]
> > > Reverted in [73b1f66](https://github.com/scylladb/scylladb/commit/73b1f66b707682c95c6a9427f00a94c5080c2f91)
> > 
> > 
> > @avikivity I'm a bit confused. How does this change affect memtable_test?
> 
> The test fails with the patch and passes with it reverted.

Evidently so, though I can't explain why exactly.
Anyhow, with https://github.com/scylladb/scylladb/pull/21299, we should be good to go with this series.
I guess I'll need to submit a new PR

**Pull request merged**
